File: articles/calculus-for-systems-modeling/articles/numerical-integration-for-systems-modeling/julia/numerical_integration_audit.jl
rate(t)=2.0+sin(t)+0.1*t
trueint(t)=2.0*t-cos(t)+1.0+0.05*t^2
h=0.1; left=0.0; trap=0.0
println("index,time,rate,left_cumulative,trapezoid_cumulative,true_cumulative,trapezoid_absolute_error")
for i in 0:100
    t=i*h; r=rate(t)
    if i>0
        global left += rate((i-1)*h)*h
        global trap += 0.5*(rate((i-1)*h)+r)*h
    end
    truth=trueint(t)-trueint(0.0)
    println(join((i,t,r,left,trap,truth,abs(trap-truth)),","))
end

File: articles/calculus-for-systems-modeling/articles/numerical-integration-for-systems-modeling/c/numerical_integration_audit.c
#include <math.h>
#include <stdio.h>
double rate(double t){ return 2.0 + sin(t) + 0.1*t; }
double trueint(double t){ return 2.0*t - cos(t) + 1.0 + 0.05*t*t; }
int main(){ double h=0.1,left=0,trap=0; printf("index,time,rate,left_cumulative,trapezoid_cumulative,true_cumulative,error\n"); for(int i=0;i<=100;i++){ double t=i*h,r=rate(t); if(i>0){ left+=rate((i-1)*h)*h; trap+=0.5*(rate((i-1)*h)+r)*h; } double truth=trueint(t)-trueint(0); printf("%d,%.6f,%.12f,%.12f,%.12f,%.12f,%.12f\n",i,t,r,left,trap,truth,fabs(trap-truth)); } return 0; }
